sklearn/preprocessing/data.py

```diff
@@ -2567,3 +2571,265 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+                    X[:, i][~valid_mask] = self.categories_[i][0]
+            X_int[:, i] = self._label_encoders_[i].transform(X[:, i])
+
+        if self.encoding == 'ordinal':
+            return X_int.astype(self.dtype, copy=False)
+
+        mask = X_mask.ravel()
+        n_values = [cats.shape[0] for cats in self.categories_]
+        n_values = np.array([0] + n_values)
+        indices = np.cumsum(n_values)
+
+        column_indices = (X_int + indices[:-1]).ravel()[mask]
+        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
+                                n_features)[mask]
+        data = np.ones(n_samples * n_features)[mask]
+
+        out = sparse.csc_matrix((data, (row_indices, column_indices)),
+                                shape=(n_samples, indices[-1]),
+                                dtype=self.dtype).tocsr()
+        if self.encoding == 'onehot-dense':
+            return out.toarray()
+        else:
+            return out
+
+    def inverse_transform(self, X):
+        """Convert back the data to the original representation.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape [n_samples, n_encoded_features]
+            The transformed data.
+
+        Returns
+        -------
+        X_tr : array-like, shape [n_samples, n_features]
+            Inverse transformed array.
+
+        """
+        check_is_fitted(self, 'categories_')
+
+        n_samples, _ = X.shape
+        n_features = len(self.categories_)
+
+        dt = np.find_common_type([cat.dtype for cat in self.categories_], [])
+        X_tr = np.empty((n_samples, n_features), dtype=dt)
+
+        if self.encoding == 'ordinal':
+            for i in range(n_features):
+                labels = X[:, i].astype('int64')
+                X_tr[:, i] = self.categories_[i][labels]
+
+        else:  # encoding == 'onehot' / 'onehot-dense'
+            j = 0
+            found_unknown = {}
+
+            for i in range(n_features):
+                n_categories = len(self.categories_[i])
+                sub = X[:, j:j + n_categories]
+
+                # for sparse X argmax returns 2D matrix, ensure 1D array
+                labels = np.asarray(sub.argmax(axis=1)).flatten()
+                X_tr[:, i] = self.categories_[i][labels]
+
+                if self.handle_unknown == 'ignore':
+                    # ignored unknown categories: we have a row of all zero's
+                    unknown = np.asarray(sub.sum(axis=1) == 0).flatten()
+                    if unknown.any():
+                        found_unknown[i] = unknown
+
+                j += n_categories
+
+            # if ignored are found: potentially need to upcast result to
+            # insert None values
+            if found_unknown:
+                if X_tr.dtype != object:
+                    X_tr = X_tr.astype(object)
+
+                for idx, mask in found_unknown.items():
+                    X_tr[mask, idx] = None
```

Comment:
I went for inserting `None` for now instead of `np.nan`. That 'feels' a bit more correct as missing value indicator (although of course in pandas we use np.nan for that, so no problem in changing this). 
One potential advantage of np.nan could be that you can presevere a numerical dtype (although int would be converted to float), but that makes deciding the final dtype somewhat more complex as it is now.

(can of course also easily change to np.nan while keeping the general object dtype as it is now)

Reply:
Hm I feel like maybe we don't want to upcast to object in this case, if it wasn't object already. Though I guess the input is unlikely to be float (what happens if someone passes floats to fit btw?).

I'm slightly leaning towards ``np.nan`` but don't have a strong opinion.

Reply:
> Hm I feel like maybe we don't want to upcast to object in this case, if it wasn't object already.

So logic then needs to be: for numeric (most likely int), cast to float64, for everything else (eg string types), cast to object. OK, not too complex to add.

Reply:
yeah... it's a bit complicated, but not overly complicated. I'm not sure if it's much better than always going to object...